code/forms/gridfield/GridFieldSiteTreeAddNewButton.php

```diff
@@ -13,34 +13,38 @@
 class GridFieldSiteTreeAddNewButton extends GridFieldAddNewButton
 	implements GridField_ActionProvider {
 
-
+	/**
+	 * Determine the list of classnames and titles allowed for a given parent object
+	 *
+	 * @param SiteTree $parent
+	 * @return boolean
+	 */
 	public function getAllowedChildren(SiteTree $parent) {
+		if(!$parent->canAddChildren()) return array();
 		$allowedChildren = $parent->allowedChildren();
-		if(empty($allowedChildren)) return array();
-
 		$children = array();
 		foreach($allowedChildren as $class) {
 			if(Config::inst()->get($class, "show_in_sitetree") === false) {
-				$children[$class] = Injector::inst()->create($class)->i18n_singular_name();
+				$instance = singleton($class);
+				// Note: Second argument to SiteTree::canCreate is proposed API
+				if($instance->canCreate(null, array('Parent' => $parent))) {

```

Comment:
More info on this? I don't think it should be merged unless its accepted in the composer.json's required version.